interfaces/dbus/backend.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package dbus implements interaction between snappy and dbus.
+//
+// Snappy creates dbus configuration files that describe how various
+// services on the system bus can communicate with other peers.
+//
+// Each configuration is an XML file containing <busconfig>...</busconfig>.
+// Particular security snippets define whole <policy>...</policy> entires.
+//
+// NOTE: This interacts with systemd.
+// TODO: Explain how this works (security).
+package dbus
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/interfaces"
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+// Backend is responsible for maintaining DBus policy files.
+type Backend struct{}
+
+// Configure creates dbus configuration files specific to a given snap.
+//
+// NOTE: DBus has no concept of a complain mode so developer mode is not supported.
+func (b *Backend) Configure(snapInfo *snap.Info, developerMode bool, repo *interfaces.Repository) error {
+	// Get the snippets that apply to this snap
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name, interfaces.SecurityDBus)
+	if err != nil {
+		return fmt.Errorf("cannot obtain security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	// Get the files that this snap should have
+	content, err := b.combineSnippets(snapInfo, developerMode, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected security files for snap %q: %s", snapInfo.Name, err)
+	}
+	glob := fmt.Sprintf("%s.conf", interfaces.SecurityTagGlob(snapInfo))
+	_, _, err = osutil.EnsureDirState(dirs.SnapBusPolicyDir, glob, content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapInfo.Name, err)
+	}
```

Comment:
I'm not 100% sure that dbus daemon doens't need to be force-reloaded. Looking at deb packaging for bluez, it has 'invoke-rc.d dbus force-reload'. On 15.04 I don't recall needing to do anything special about this, but it might've been because we were using Type=dbus and BusName=... in the systemd unit. Also, Tony didn't complain about the bus policy not taking effect immediately with his earlier 16.04 snap.... I looked at the dbus code and it has an inotify watch on configuration and service directories.

In all, I don't think there is anything to do here, but I wanted to add a comment here for future reference.